File: ios-companion/GuardiaCompanion/ApiClient.swift
import Foundation

/// Minimal JSON-over-HTTPS client — mirrors the Android companion's
/// ApiClient.kt. No third-party networking library needed for two POSTs.
enum ApiClient {
    struct PairResult {
        let token: String
        let studentName: String
    }

    static func pairDevice(pairCode: String, deviceName: String, completion: @escaping (PairResult?) -> Void) {
        let body: [String: Any] = ["pairCode": pairCode, "deviceName": deviceName]
        post(path: "/api/device/pair", body: body) { json in
            guard let json, let token = json["token"] as? String else {
                completion(nil)
                return
            }
            let studentName = json["studentName"] as? String ?? "your child"
            completion(PairResult(token: token, studentName: studentName))
        }
    }

    static func ingest(token: String, packageName: String, role: String, text: String, completion: @escaping (Bool) -> Void) {
        let body: [String: Any] = ["token": token, "packageName": packageName, "role": role, "text": text]
        post(path: "/api/device/ingest", body: body) { json in
            completion((json?["ok"] as? Bool) ?? false)
        }
    }

    /// Reports a Screen Time threshold crossing — a count of minutes, never
    /// message content. `minutes` is today's threshold that was just
    /// reached (e.g. 30), not a live running total; see SharedUsageStore.
    static func reportUsage(token: String, packageName: String, minutes: Int, completion: @escaping (Bool) -> Void) {
        let body: [String: Any] = ["token": token, "packageName": packageName, "minutes": minutes]
        post(path: "/api/device/usage", body: body) { json in
            completion((json?["ok"] as? Bool) ?? false)
        }
    }

    /// Drains whatever threshold events the extension recorded while this
    /// app wasn't running, and reports each one. Call from `.onAppear` /
    /// `scenePhase == .active` in the main app.
    static func syncPendingUsageEvents() {
        guard let token = Prefs.token else { return }
        let pending = SharedUsageStore.readPending()
        guard !pending.isEmpty else { return }
        SharedUsageStore.clearPending() // clear first: a duplicate re-send is harmless (see route's comment), a lost event is not

        for event in pending {
            reportUsage(token: token, packageName: event.packageName, minutes: event.minutes) { _ in }
        }
    }

    private static func post(path: String, body: [String: Any], completion: @escaping ([String: Any]?) -> Void) {
        guard let url = URL(string: Prefs.baseURL + path) else {
            completion(nil)
            return
        }
        var request = URLRequest(url: url, timeoutInterval: 8)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try? JSONSerialization.data(withJSONObject: body)

        URLSession.shared.dataTask(with: request) { data, response, error in
            guard error == nil, let data,
                  let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode),
                  let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
            else {
                DispatchQueue.main.async { completion(nil) }
                return
            }
            DispatchQueue.main.async { completion(json) }
        }.resume()
    }
}

File: mobile-companion/src/api/ApiClient.ts
import { Platform } from 'react-native';

/** Mirrors the native apps' ApiClient (ios-companion/ApiClient.swift,
 *  android-companion/ApiClient.kt) — same three endpoints, same shapes.
 *  Android's emulator needs the 10.0.2.2 host alias for the dev machine's
 *  localhost; iOS's simulator shares the Mac's own network stack so
 *  localhost works directly. Release builds always use the real deployment. */
const DEV_BASE_URL = Platform.select({
  android: 'http://10.0.2.2:3912',
  default: 'http://localhost:3912',
});
const PROD_BASE_URL = 'https://guardia-seven.vercel.app';

export const BASE_URL = __DEV__ ? DEV_BASE_URL : PROD_BASE_URL;

async function post<T>(path: string, body: Record<string, unknown>): Promise<T | null> {
  try {
    const res = await fetch(`${BASE_URL}${path}`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
    });
    if (!res.ok) return null;
    return (await res.json()) as T;
  } catch {
    return null;
  }
}

export type PairResult = { token: string; studentName: string };

export const ApiClient = {
  async pairDevice(pairCode: string, deviceName: string): Promise<PairResult | null> {
    const json = await post<{ token?: string; studentName?: string }>('/api/device/pair', {
      pairCode,
      deviceName,
    });
    if (!json?.token) return null;
    return { token: json.token, studentName: json.studentName ?? 'your child' };
  },

  async ingest(token: string, packageName: string, role: string, text: string): Promise<boolean> {
    const json = await post<{ ok?: boolean }>('/api/device/ingest', { token, packageName, role, text });
    return json?.ok ?? false;
  },

  /** Reports a Screen Time threshold crossing — a minute count, never
   *  message content. `minutes` is the threshold just reached (e.g. 30),
   *  not a live running total. */
  async reportUsage(token: string, packageName: string, minutes: number): Promise<boolean> {
    const json = await post<{ ok?: boolean }>('/api/device/usage', { token, packageName, minutes });
    return json?.ok ?? false;
  },
};
